src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -3504,6 +3505,54 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
+        {
+            BoundExpression leftOperand = BindValue(node.Left, diagnostics, BindValueKind.CompoundAssignment);
+            BoundExpression rightOperand = BindValue(node.Right, diagnostics, BindValueKind.RValue);
+
+            // If either operand's Type is dynamic, then we potentially need to check overflow at runtime
+            bool isChecked = leftOperand.HasDynamicType() || rightOperand.HasDynamicType() ? CheckOverflowAtCompileTime : false;
+
+            // If either operand is bad, bail out preventing more cascading errors
+            if (leftOperand.HasAnyErrors || rightOperand.HasAnyErrors)
+            {
+                return new BoundNullCoalescingAssignmentOperator(node, leftOperand, rightOperand, isChecked, CreateErrorType(), hasErrors: true);
+            }
+
+            // Unlike a standard null coalescing expression, the resulting type of the expression a ??= b
+            // must be A (where A is the type of a), as it is where the result of the assignment will end up. Therefore,
+            // we must ensure that B (where B is the type B) is implicitly convertible to A or A0 (where A0 is the
+            // underlying non-nullable type of A).
+            TypeSymbol leftType = leftOperand.Type;
+            Debug.Assert(leftType != null);
+
+            // If A is not a nullable type or reference type, a compile-time error occurs
+            if (!leftType.IsReferenceType && !leftType.IsNullableType())
+            {
+                return GenerateNullCoalescingAssignmentBadBinaryOpsError(node, leftOperand, rightOperand, isChecked, diagnostics);
+            }
+
+            // If an implicit conversion exists from B to A, we store that conversion. At runtime, a is first evaluated. If
+            // a is not null, b is not evaluated. If a is null, b is evaluated and converted to type A, and is stored in a.
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            var rightConversion = Conversions.ClassifyImplicitConversionFromExpression(rightOperand, leftType, ref useSiteDiagnostics);
+            diagnostics.Add(node, useSiteDiagnostics);
+            if (rightConversion.Exists)
+            {
+                var convertedRightOperand = CreateConversion(rightOperand, rightConversion, leftType, diagnostics);
+                return new BoundNullCoalescingAssignmentOperator(node, leftOperand, convertedRightOperand, isChecked, leftType);
+            }
+
+            // a and b are incompatible and a compile-time error occurs
+            return GenerateNullCoalescingAssignmentBadBinaryOpsError(node, leftOperand, rightOperand, isChecked, diagnostics);
+        }
+
+        private BoundExpression GenerateNullCoalescingAssignmentBadBinaryOpsError(AssignmentExpressionSyntax node, BoundExpression leftOperand, BoundExpression rightOperand, bool isChecked, DiagnosticBag diagnostics)
```

Comment:
Consider making this a local function, since only used in one method.

Reply:
@cston I can't win!

---
In reply to: [201188465](https://github.com/dotnet/roslyn/pull/28253#discussion_r201188465) [](ancestors = 201188465)